subsys/net/ip/igmp.c

```diff
@@ -56,14 +70,119 @@ static int igmp_v2_create(struct net_pkt *pkt, const struct in_addr *addr,
 	igmp->max_rsp = 0U;
 	net_ipaddr_copy(&igmp->address, addr);
 	igmp->chksum = 0;
-	igmp->chksum = net_calc_chksum_igmp((uint8_t *)igmp, sizeof(*igmp));
 
 	if (net_pkt_set_data(pkt, &igmp_access)) {
 		return -ENOBUFS;
 	}
 
+	igmp->chksum = net_calc_chksum_igmp(pkt);
+
+	net_pkt_set_overwrite(pkt, true);
+	net_pkt_cursor_init(pkt);
+
+	net_pkt_skip(pkt, offsetof(struct net_ipv4_igmp_v2_report, chksum));
+	if (net_pkt_write(pkt, &igmp->chksum, sizeof(igmp->chksum))) {
+		return -ENOBUFS;
+	}
+
+	return 0;
+}
+
+#if defined(CONFIG_NET_IPV4_IGMPV3)
+static int igmp_v3_create(struct net_pkt *pkt, uint8_t type, struct net_if_mcast_addr mcast[],
+			  size_t mcast_len)
+{
+	NET_PKT_DATA_ACCESS_DEFINE(igmp_access, struct net_ipv4_igmp_v3_report);
+	NET_PKT_DATA_ACCESS_DEFINE(group_record_access, struct net_ipv4_igmp_v3_group_record);
+	struct net_ipv4_igmp_v3_report *igmp;
+	struct net_ipv4_igmp_v3_group_record *group_record;
+
+	uint16_t group_count = 0;
+	const struct in_addr all_systems = ALL_SYSTEMS_ADDR;
+
+	igmp = (struct net_ipv4_igmp_v3_report *)net_pkt_get_data(pkt, &igmp_access);
+	if (!igmp) {
+		return -ENOBUFS;
+	}
+
+	for (int i = 0; i < mcast_len; i++) {
+		/* We don't need to send an IGMP membership report to the IGMP
+		 * all systems multicast address of 224.0.0.1 so skip over it.
+		 * Since the IGMP all systems multicast address is marked as
+		 * used and joined during init time, we have to check this
+		 * address separately to skip over it.
+		 */
+		if (!mcast[i].is_used || !mcast[i].is_joined ||
+		    net_ipv4_addr_cmp_raw((uint8_t *)&mcast[i].address.in_addr,
+					  (uint8_t *)&all_systems)) {
+			continue;
+		}
+
+		group_count++;
+	}
+
+	igmp->type = type;
+	igmp->reserved_1 = 0U;
+	igmp->reserved_2 = 0U;
+	igmp->groups_len = htons(group_count);
+	/* Setting initial value of chksum to 0 to calculate chksum as described in RFC 3376
+	 * ch 4.1.2
+	 */
+	igmp->chksum = 0;
+
+	if (net_pkt_set_data(pkt, &igmp_access)) {
+		return -ENOBUFS;
+	}
+
+	for (int i = 0; i < NET_IF_MAX_IPV4_MADDR; i++) {
```

Comment:
Shouldn't this iterate until `i < mcast_len` instead? `igmpv3_send_generic()` will only provide a single address to put into the report.

Reply:
Indeed